Fix platform switching

bhollis opened the pull request:
Things had gotten a bit byzantine around platform switching and updating stores. This attempts to fix it. In the process I reverted some of the fixes we've applied recently. Long story short, 2253c8c added a bug that unassigned `_stores` while stores were updating, and then 1135831511936ddcd182f8342d313d67e8e57fbe and d71b308161954364fe682d8ab6d7d867169639b8 were added in order to work around that bug - but what was actually needed was to simplify things and stop using "platform updated" to trigger refreshes. This PR untangles all that stuff. It also gets rid of hacks like `$rootScope.activePlatformUpdated` which are no longer necessary, and properly fires events so that all items disappear right when you change platform, etc.

As a consequence of the untangling we end up avoiding several service calls that we were doing on every refresh (and a handful elsewhere) which improves DIM performance.

@SunburnedGoose, I'd appreciate if you could test this out - it works for everything I can test, but I don't have an Xbox so I can't test all the double platform stuff. 

/cc @ericnelson0

SunburnedGoose commented:
Shiny and Chrome...  This bug is back.

![screen shot 2016-07-22 at 1 26 46 am](https://cloud.githubusercontent.com/assets/984608/17048416/6484fd22-4fab-11e6-9a9c-fdb71f6a0f5d.png)

bhollis commented:
@SunburnedGoose you'll have to explain a bit more what "this bug" is - I can't figure it out from the screenshot.

SunburnedGoose commented:
Edited the message because it was vague.  When you switch platforms, all the items are marked as new.

bhollis commented:
Note to self: might as well fix #698 while I'm in here.

bhollis commented:
Ah. I don't think your screenshot is animated. 😉 

I'll look into it. I think the "new item" stuff needs a minor overhaul anyway. Linking in #711 which is easier for me to reproduce (unless somebody wants to send me an Xbox One (don't really do that)).

bhollis commented:
I popped in what should be a quick fix for the new-item thing, but I'm still going to spend some time really understanding that code and shoring it up a bit more.

ericnelson0 commented:
@bhollis LMK if you have any questions on that code.

BTW, on #711 an easy fix might be to look at why ng-disabled on those links isn't working while the stores are being loaded.

bhollis commented:
@ericnelson0 `ng-disabled` doesn't disable links, or disable `ng-click`...

bhollis commented:
@SunburnedGoose does this fix the new-items-on-platform-change bug? Any other bugs you see?

bhollis commented:
OK, I've gone pretty deep here. Some new changes:

1. You can no longer accidentally trigger two `reloadStores` at the same time. Helpful for preventing extra work and concurrent mutated global state.
1. Rewrote the new-item code. It shares less global state, properly maintains "new-ness" in sync service across reloads/browsers, and stores new-ness per-platform. The popup notification now only notifies about stuff that's new since the last refresh, as opposed to popping up a cumulative list every time.
2. Added a "clear new items" button. This fixes #628. The keyboard shortcut is "x".
3. Enabled the automatic cheat sheet for keyboard shortcuts. (It's shown with "?")

![screen shot 2016-07-23 at 7 07 34 pm](https://cloud.githubusercontent.com/assets/313208/17081331/cdfc9310-5108-11e6-9471-1d70f0d9b757.png)
![screen shot 2016-07-23 at 7 07 51 pm](https://cloud.githubusercontent.com/assets/313208/17081332/d0484998-5108-11e6-819a-74830ab0d053.png)

Mulchman commented:
Will love the 'x' button addition!

On Jul 23, 2016 19:10, "Ben Hollis" <notifications@github.com> wrote:

> OK, I've gone pretty deep here. Some new changes:
>
>    1. You can no longer accidentally trigger two reloadStores at the same
>    time. Helpful for preventing extra work and concurrent mutated global state.
>    2. Rewrote the new-item code. It shares less global state, properly
>    maintains "new-ness" in sync service across reloads/browsers, and stores
>    new-ness per-platform. The popup notification now only notifies about stuff
>    that's new since the last refresh, as opposed to popping up a cumulative
>    list every time.
>    3. Added a "clear new items" button. This fixes #628
>    <https://github.com/DestinyItemManager/DIM/issues/628>. The keyboard
>    shortcut is "x".
>    4. Enabled the automatic cheat sheet for keyboard shortcuts. (It's
>    shown with "?")
>
> [image: screen shot 2016-07-23 at 7 07 34 pm]
> <https://cloud.githubusercontent.com/assets/313208/17081331/cdfc9310-5108-11e6-9471-1d70f0d9b757.png>
> [image: screen shot 2016-07-23 at 7 07 51 pm]
> <https://cloud.githubusercontent.com/assets/313208/17081332/d0484998-5108-11e6-819a-74830ab0d053.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DestinyItemManager/DIM/pull/713#issuecomment-234752430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALNpqVaRTSrUBawtC620260vBpZ2h4J3ks5qYsmlgaJpZM4JSeLz>
> .
>

kyleshay commented:
keyboard cheatsheet is really nice looking!! maybe consider linking to it in the about or settings dialog?

bhollis commented:
@kyleshay good idea! I added something to the FAQ.

There's enough stuff fixed in this PR that I'd like to just merge it, and if we still have bugs around platform switching or anything else, we can file another issue for it.
